include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,230 @@
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    iterator() = default;
+
+    iterator(iterator const&) = default;
+
+    iterator&
+    operator=(iterator const&) = default;
+
+    BOOST_URL_DECL
+    reference
+    operator*() const noexcept;
+
+    iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        if (*pos_ != '%')
+            ++pos_;
+        else
+            pos_ += 3;
+        return *this;
+    }
+
+    iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ < 3 ||
+                pos_[-3] != '%')
+            --pos_;
+        else
+            pos_ -= 3;
+        return *this;
+    }
+
+    iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    bool
+    operator==(
+        iterator const& other) const noexcept
+    {
+        return pos_ == other.pos_;
+    }
+
+    bool
+    operator!=(
+        iterator const& other) const noexcept
+    {
+        return !(*this == other);
+    }
+};
+
+inline
+auto
+decoded_view::
+begin() const noexcept ->
+    const_iterator
+{
+    return {p_, n_, plus_to_space_};
+}
+
+inline
+auto
+decoded_view::
+end() const noexcept ->
+    const_iterator
+{
+    return {p_, n_, plus_to_space_, 0};
+}
+
+inline
+auto
+decoded_view::
+front() const noexcept ->
+    const_reference
+{
+    BOOST_ASSERT( !empty() );
+    return *begin();
+}
+
+inline
+auto
+decoded_view::
+back() const noexcept ->
+    const_reference
+{
+    BOOST_ASSERT( !empty() );
+    return *--end();
+}
+
+namespace detail
+{
+template<class C, class I, class = void>
+struct has_assign_from : std::false_type {};
+
+template<class C, class I>
+struct has_assign_from<C, I, boost::void_t<
+    decltype(
+        std::declval<C&>()
+            .assign(
+                std::declval<I>(),
+                std::declval<I>())
+                ) > > : std::true_type
+{
+};
+
+template<class C, class I, class = void>
+struct has_append_from : std::false_type {};
+
+template<class C, class I>
+struct has_append_from<C, I, boost::void_t<
+    decltype(
+        std::declval<C&>()
+            .append(
+                std::declval<I>(),
+                std::declval<I>())
+                ) > > : std::true_type
+{
+};
+}
+
+template <class MutableString>
+MutableString&
+decoded_view::
+assign_to(
+    MutableString& s) const
+{
+    // If this goes off, it means that your String type
```

Comment:
"If this goes off" is what you write for an assert. For a static assert, start the sentence with "If you get a compile error here"